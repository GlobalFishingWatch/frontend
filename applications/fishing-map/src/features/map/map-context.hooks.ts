import { useContext } from 'react'
import { MapContext, ViewportProps, WebMercatorViewport } from 'react-map-gl'
import type { Map } from '@globalfishingwatch/mapbox-gl'

export type ViewportChangeHandler = (viewState: ViewportProps) => void
export interface ExtraState {
  inTransition?: boolean | undefined
  isDragging?: boolean | undefined
  isHovering?: boolean | undefined
  isPanning?: boolean | undefined
  isRotating?: boolean | undefined
  isZooming?: boolean | undefined
}
export type ContextViewportChangeHandler = (
  viewState: ViewportProps,
  interactionState: ExtraState,
  oldViewState: ViewportProps
) => void

export interface ViewStateChangeInfo {
  viewState: ViewportProps
}

export type ViewStateChangeHandler = (info: ViewStateChangeInfo) => void

export type ForkedContext = {
  viewport?: WebMercatorViewport
  map?: Map // need to overwrite types here to use '@globalfishingwatch/mapbox-gl' Map type
  container: HTMLElement | null
  mapContainer: HTMLElement | null
  onViewStateChange?: ViewStateChangeHandler
  onViewportChange?: ContextViewportChangeHandler
  isDragging: boolean
  eventManager?: any
}

// const MapContext = _MapContext as React.Context<ForkedContext>
export { MapContext }

export function useMapContext() {
  const context = useContext(MapContext)
  return context as ForkedContext
}

export default function useMapInstance() {
  const { map } = useMapContext()
  return map
}
