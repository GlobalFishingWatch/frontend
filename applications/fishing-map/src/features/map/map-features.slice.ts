import { createSlice, PayloadAction } from '@reduxjs/toolkit'
import { RootState } from 'store'
import { InteractionEvent } from '@globalfishingwatch/react-hooks'

type MapFeaturesState = {
  clicked: InteractionEvent | null
  hovered: InteractionEvent | null
}

const initialState: MapFeaturesState = {
  clicked: null,
  hovered: null,
}

const slice = createSlice({
  name: 'mapFeatures',
  initialState,
  reducers: {
    setClickedEvent: (state, action: PayloadAction<InteractionEvent | null>) => {
      state.clicked = action.payload
    },
  },
})

export const selectClickedFeatures = (state: RootState) => state.mapFeatures.clicked

export const { setClickedEvent } = slice.actions
export default slice.reducer
