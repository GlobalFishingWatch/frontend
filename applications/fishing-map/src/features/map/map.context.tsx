import React, { createContext, useRef, useContext, useState, useEffect, useMemo } from 'react'
import type { Map } from '@globalfishingwatch/mapbox-gl'

type MapStateContextPayload = {
  mapboxRef?: any
  mapboxInstance?: Map
}

const MapStateContext = createContext<MapStateContextPayload>({})

const MapboxRefProvider: React.FC = ({ children }) => {
  const mapboxRef = useRef()
  const [mapboxRefReady, seMapboxRefReady] = useState(false)
  const [mapboxInstance, seMapboxInstance] = useState<Map>()
  // eslint-disable-next-line react-hooks/exhaustive-deps
  useEffect(() => {
    if (mapboxRef.current !== undefined && !mapboxRefReady) {
      seMapboxRefReady(true)
      seMapboxInstance((mapboxRef.current as any).getMap())
    }
  })
  const context = useMemo(
    () => ({
      mapboxRef,
      mapboxInstance,
    }),
    [mapboxInstance]
  )
  return <MapStateContext.Provider value={context}>{children}</MapStateContext.Provider>
}

function useMapboxRef() {
  const context = useContext(MapStateContext)
  if (context === undefined) {
    throw new Error('useMapboxRef must be used within a MapboxRefProvider')
  }
  return context.mapboxRef
}

function useMapboxInstance() {
  const context = useContext(MapStateContext)
  if (context === undefined) {
    throw new Error('useMapboxInstance must be used within a MapboxRefProvider')
  }
  return context.mapboxInstance
}

export { MapboxRefProvider, useMapboxRef, useMapboxInstance }
