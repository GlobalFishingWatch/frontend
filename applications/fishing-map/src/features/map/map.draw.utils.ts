import {
  Dataset,
  DatasetCategory,
  DatasetConfiguration,
  DatasetTypes,
} from '@globalfishingwatch/api-types'
import { DrawFeature, DrawPointPosition } from './MapDraw'

export const getDrawDatasetDefinition = (name: string): Partial<Dataset> => {
  return {
    name,
    type: DatasetTypes.Context,
    category: DatasetCategory.Context,
    configuration: {
      format: 'geojson',
      geometryType: 'draw',
    } as DatasetConfiguration,
  }
}

export const getFileWithFeatures = (name: string, features: DrawFeature[]) => {
  return new File(
    [
      JSON.stringify({
        type: 'FeatureCollection',
        features: features,
      }),
    ],
    `${name}.json`,
    {
      type: 'application/json',
    }
  )
}

export const getFeaturesPrecisionRounded = (features: DrawFeature[]): DrawFeature[] => {
  return features.map(
    (feature) =>
      ({
        ...feature,
        geometry: {
          ...feature.geometry,
          coordinates: feature.geometry.coordinates.map((coordinate) =>
            coordinate.map((points) => points.map((point) => Math.round(point * 100000) / 100000))
          ),
        },
      } as DrawFeature)
  )
}

export const updateFeaturePointByIndex = (
  features: DrawFeature[],
  featureIndex: number | undefined,
  coordinateIndex: number | undefined,
  pointPosition: DrawPointPosition
): DrawFeature[] => {
  if (featureIndex === undefined || coordinateIndex === undefined) {
    return features
  }
  return features?.map((feature, index) => {
    if (index !== featureIndex) {
      return feature
    }
    return {
      ...feature,
      geometry: {
        ...feature.geometry,
        coordinates: feature.geometry.coordinates.map((coordinates) => {
          const coordinatesLength = coordinates.length - 1
          const isFirstCoordinate = coordinateIndex === 0
          const isLastCoordinate = coordinateIndex === coordinatesLength
          return coordinates.map((point, index) => {
            if (
              index === coordinateIndex ||
              (isLastCoordinate && index === 0) ||
              (isFirstCoordinate && index === coordinatesLength)
            ) {
              return pointPosition
            }
            return point
          })
        }),
      },
    } as DrawFeature
  })
}
