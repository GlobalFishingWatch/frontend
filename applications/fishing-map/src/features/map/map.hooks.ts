import { useSelector, useDispatch } from 'react-redux'
import { Geometry } from 'geojson'
import { useCallback, useEffect, useMemo, useRef } from 'react'
import { debounce } from 'lodash'
import { InteractionEvent } from '@globalfishingwatch/react-hooks'
import { Generators } from '@globalfishingwatch/layer-composer'
import {
  MULTILAYER_SEPARATOR,
  MERGED_ACTIVITY_ANIMATED_HEATMAP_GENERATOR_ID,
  UrlDataviewInstance,
} from '@globalfishingwatch/dataviews-client'
import { ContextLayerType, Type } from '@globalfishingwatch/layer-composer/dist/generators/types'
import type { Style } from '@globalfishingwatch/mapbox-gl'
import { DataviewCategory } from '@globalfishingwatch/api-types/dist'
import {
  TemporalGridFeature,
  useFeatureState,
} from '@globalfishingwatch/react-hooks/dist/use-map-interaction'
import GFWAPI from '@globalfishingwatch/api-client'
import { ENCOUNTER_EVENTS_SOURCE_ID } from 'features/dataviews/dataviews.utils'
import { selectLocationType } from 'routes/routes.selectors'
import { HOME, USER, WORKSPACE, WORKSPACES_LIST } from 'routes/routes'
import { useLocationConnect } from 'routes/routes.hook'
import { DEFAULT_WORKSPACE_ID, WorkspaceCategories } from 'data/workspaces'
import useMapInstance from 'features/map/map-context.hooks'
import { removeDatasetVersion } from 'features/datasets/datasets.utils'
import { useTimerangeConnect } from 'features/timebar/timebar.hooks'
import { selectHighlightedEvent, setHighlightedEvent } from 'features/timebar/timebar.slice'
import { t } from 'features/i18n/i18n'
import {
  selectDefaultMapGeneratorsConfig,
  WORKSPACES_POINTS_TYPE,
  WORKSPACE_GENERATOR_ID,
} from './map.selectors'
import {
  setClickedEvent,
  selectClickedEvent,
  MAX_TOOLTIP_LIST,
  fetchEncounterEventThunk,
  SliceInteractionEvent,
  selectFishingInteractionStatus,
  selectApiEventStatus,
  ExtendedFeatureVessel,
  ExtendedFeatureEvent,
  fetchFishingActivityInteractionThunk,
  fetchViirsInteractionThunk,
  selectViirsInteractionStatus,
  ApiViirsStats,
} from './map.slice'
import useViewport from './map-viewport.hooks'
import { useMapAndSourcesLoaded, useMapLoaded } from './map-features.hooks'

// This is a convenience hook that returns at the same time the portions of the store we interested in
// as well as the functions we need to update the same portions
export const useGeneratorsConnect = () => {
  const { start, end } = useTimerangeConnect()
  const { viewport } = useViewport()
  const generatorsConfig = useSelector(selectDefaultMapGeneratorsConfig)

  return useMemo(() => {
    return {
      generatorsConfig,
      globalConfig: {
        zoom: viewport.zoom,
        start,
        end,
        token: GFWAPI.getToken(),
      },
    }
  }, [generatorsConfig, viewport.zoom, start, end])
}

export const useClickedEventConnect = () => {
  const map = useMapInstance()
  const dispatch = useDispatch()
  const clickedEvent = useSelector(selectClickedEvent)
  const locationType = useSelector(selectLocationType)
  const fishingInteractionStatus = useSelector(selectFishingInteractionStatus)
  const viirsInteractionStatus = useSelector(selectViirsInteractionStatus)
  const apiEventStatus = useSelector(selectApiEventStatus)
  const { dispatchLocation } = useLocationConnect()
  const { cleanFeatureState } = useFeatureState(map)
  const { setMapCoordinates } = useViewport()
  const encounterSourceLoaded = useMapAndSourcesLoaded(ENCOUNTER_EVENTS_SOURCE_ID)
  const fishingPromiseRef = useRef<any>()
  const viirsPromiseRef = useRef<any>()
  const eventsPromiseRef = useRef<any>()

  const dispatchClickedEvent = (event: InteractionEvent | null) => {
    if (event === null) {
      dispatch(setClickedEvent(null))
      return
    }
    // Used on workspaces-list or user panel to go to the workspace detail page
    if (locationType === USER || locationType === WORKSPACES_LIST) {
      const workspace = event?.features?.find(
        (feature: any) => feature.properties.type === WORKSPACES_POINTS_TYPE
      )
      if (workspace) {
        const isDefaultWorkspace = workspace.properties.id === DEFAULT_WORKSPACE_ID
        dispatchLocation(
          isDefaultWorkspace ? HOME : WORKSPACE,
          isDefaultWorkspace
            ? {}
            : {
                category:
                  workspace.properties?.category && workspace.properties.category !== 'null'
                    ? workspace.properties.category
                    : WorkspaceCategories.FishingActivity,
                workspaceId: workspace.properties.id,
              },
          true
        )
        const { latitude, longitude, zoom } = workspace.properties
        if (latitude && longitude && zoom) {
          setMapCoordinates({ latitude, longitude, zoom })
        }
        return
      }
    }

    const clusterFeature = event?.features?.find(
      (f) => f.generatorType === Generators.Type.TileCluster
    )
    if (clusterFeature?.properties?.expansionZoom) {
      const { count, expansionZoom, lat, lng } = clusterFeature.properties
      if (count > 1) {
        if (encounterSourceLoaded) {
          setMapCoordinates({
            latitude: lat,
            longitude: lng,
            zoom: expansionZoom,
          })
          cleanFeatureState('click')
        }
        return
      }
    }

    // Cancel all pending promises
    const promisesRef = [fishingPromiseRef, viirsPromiseRef, eventsPromiseRef]
    promisesRef.forEach((ref) => {
      if (ref.current) {
        ref.current.abort()
      }
    })

    if (!event || !event.features) {
      if (clickedEvent) {
        dispatch(setClickedEvent(null))
      }
      return
    }

    // When hovering in a vessel event we don't want to have clicked events
    const isVesselSingleFeatureEvent =
      event.features.find((f) => f.generatorType === Generators.Type.VesselEvents) !== undefined

    if (isVesselSingleFeatureEvent && event.features?.length === 1) {
      return
    }

    dispatch(setClickedEvent(event as SliceInteractionEvent))

    // get temporal grid clicked features and order them by sublayerindex
    const fishingActivityFeatures = event.features
      .filter((feature) => {
        if (!feature.temporalgrid) {
          return false
        }
        const isFeatureVisible = feature.temporalgrid.visible
        const isFishingFeature = feature.temporalgrid.sublayerInteractionType === 'fishing-effort'
        return isFeatureVisible && isFishingFeature
      })
      .sort((feature) => feature.temporalgrid?.sublayerIndex ?? 0)

    if (fishingActivityFeatures?.length) {
      fishingPromiseRef.current = dispatch(
        fetchFishingActivityInteractionThunk({ fishingActivityFeatures })
      )
    }

    const viirsFeature = event.features?.find((feature) => {
      if (!feature.temporalgrid) {
        return false
      }
      const isFeatureVisible = feature.temporalgrid.visible
      const isViirsFeature = feature.temporalgrid.sublayerInteractionType === 'viirs'
      return isFeatureVisible && isViirsFeature
    })

    if (viirsFeature) {
      viirsPromiseRef.current = dispatch(fetchViirsInteractionThunk({ feature: viirsFeature }))
    }

    const encounterFeature = event.features.find(
      (f) => f.generatorType === Generators.Type.TileCluster
    )
    if (encounterFeature) {
      eventsPromiseRef.current = dispatch(fetchEncounterEventThunk(encounterFeature))
    }
  }
  return {
    clickedEvent,
    fishingInteractionStatus,
    viirsInteractionStatus,
    apiEventStatus,
    dispatchClickedEvent,
  }
}

// TODO this could extend ExtendedFeature
export type TooltipEventFeature = {
  id?: string
  title?: string
  type?: Type
  color?: string
  unit?: string
  source: string
  sourceLayer: string
  layerId: string
  generatorContextLayer?: ContextLayerType | null
  geometry?: Geometry
  value: string // TODO Why not a number?
  properties: Record<string, string>
  vesselsInfo?: {
    overflow: boolean
    numVessels: number
    vessels: ExtendedFeatureVessel[]
  }
  event?: ExtendedFeatureEvent
  viirs?: ApiViirsStats[]
  temporalgrid?: TemporalGridFeature
  category: DataviewCategory
}

export type TooltipEvent = {
  latitude: number
  longitude: number
  features: TooltipEventFeature[]
}

export const useMapHighlightedEvent = (features?: TooltipEventFeature[]) => {
  const highlightedEvent = useSelector(selectHighlightedEvent)
  const dispatch = useDispatch()

  // eslint-disable-next-line react-hooks/exhaustive-deps
  const debounceDispatch = useCallback(
    debounce((event: any) => {
      dispatch(setHighlightedEvent(event))
    }, 100),
    []
  )
  // eslint-disable-next-line react-hooks/exhaustive-deps
  const setHighlightedEventDebounced = useCallback(() => {
    const vesselFeature = features?.find((f) => f.category === DataviewCategory.Vessels)
    if (vesselFeature) {
      if (vesselFeature.properties?.id && highlightedEvent?.id !== vesselFeature.properties.id) {
        debounceDispatch({ id: vesselFeature.properties.id })
      }
    } else if (highlightedEvent) {
      debounceDispatch(undefined)
    }
  }, [features, highlightedEvent, debounceDispatch])

  useEffect(() => {
    setHighlightedEventDebounced()
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [features])
}

export const parseMapTooltipEvent = (
  event: SliceInteractionEvent | null,
  dataviews: UrlDataviewInstance<Generators.Type>[],
  temporalgridDataviews: UrlDataviewInstance<Generators.Type>[]
) => {
  if (!event || !event.features) return null

  const clusterFeature = event.features.find(
    (f) => f.generatorType === Generators.Type.TileCluster && parseInt(f.properties.count) > 1
  )

  // We don't want to show anything else when hovering a cluster point
  if (clusterFeature) {
    return {
      point: event.point,
      latitude: event.latitude,
      longitude: event.longitude,
      features: [
        {
          type: clusterFeature.generatorType,
          properties: clusterFeature.properties,
        } as TooltipEventFeature,
      ],
    }
  }

  const tooltipEventFeatures: TooltipEventFeature[] = event.features.flatMap((feature) => {
    let dataview
    if (feature.generatorId === MERGED_ACTIVITY_ANIMATED_HEATMAP_GENERATOR_ID) {
      const { temporalgrid } = feature
      if (!temporalgrid || temporalgrid.sublayerId === undefined || !temporalgrid.visible) {
        return []
      }

      dataview = temporalgridDataviews?.find((dataview) => dataview.id === temporalgrid.sublayerId)
    } else {
      dataview = dataviews?.find((dataview) => {
        // Needed to get only the initial part to support multiple generator
        // from the same dataview, see map.selectors L137
        const cleanGeneratorId = (feature.generatorId as string)?.split(MULTILAYER_SEPARATOR)[0]
        return dataview.id === cleanGeneratorId
      })
    }

    if (!dataview) {
      // Not needed to create a dataview just for the workspaces list interaction
      if (feature.generatorId && (feature.generatorId as string).includes(WORKSPACE_GENERATOR_ID)) {
        const tooltipWorkspaceFeature: TooltipEventFeature = {
          source: feature.source,
          sourceLayer: feature.sourceLayer,
          layerId: feature.layerId as string,
          type: Generators.Type.GL,
          value: feature.properties.label,
          properties: {},
          category: DataviewCategory.Context,
        }
        return tooltipWorkspaceFeature
      }
      return []
    }

    let title = dataview.name || dataview.id.toString()
    if (
      dataview.category === DataviewCategory.Context ||
      dataview.category === DataviewCategory.Events
    ) {
      const dataset = dataview.datasets?.[0]
      if (dataset) {
        if (dataview.config?.type === Generators.Type.UserContext) {
          title = dataset.name
        } else {
          const datasetId = removeDatasetVersion(dataset?.id)
          title = t(`datasets:${datasetId}.name` as any) as string
        }
      }
    }
    const tooltipEventFeature: TooltipEventFeature = {
      title,
      type: dataview.config?.type,
      color: dataview.config?.color || 'black',
      category: dataview.category || DataviewCategory.Context,
      ...feature,
      properties: { ...feature.properties },
    }
    // Insert custom properties by each dataview configuration
    const properties = dataview.datasetsConfig
      ? dataview.datasetsConfig.flatMap((datasetConfig) => {
          if (!datasetConfig.query?.length) return []
          return datasetConfig.query.flatMap((query) =>
            query.id === 'properties' ? (query.value as string) : []
          )
        })
      : []
    properties.forEach((property) => {
      if (feature.properties[property]) {
        tooltipEventFeature.properties[property] = feature.properties[property]
      }
    })

    if (feature.vessels) {
      tooltipEventFeature.vesselsInfo = {
        vessels: feature.vessels.slice(0, MAX_TOOLTIP_LIST),
        numVessels: feature.vessels.length,
        overflow: feature.vessels.length > MAX_TOOLTIP_LIST,
      }
    }
    return tooltipEventFeature
  })

  if (!tooltipEventFeatures.length) return null
  return {
    point: event.point,
    latitude: event.latitude,
    longitude: event.longitude,
    features: tooltipEventFeatures,
  }
}

export const useMapStyle = () => {
  const map = useMapInstance()

  // Used to ensure the style is refreshed on load finish
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  const mapLoaded = useMapLoaded()

  if (!map) return null

  let style: Style
  try {
    style = map.getStyle()
  } catch (e) {
    return null
  }

  return style
}

export const useGeneratorStyleMetadata = (generatorId: string) => {
  const style = useMapStyle()
  return style?.metadata?.generatorsMetadata[generatorId] || {}
}
