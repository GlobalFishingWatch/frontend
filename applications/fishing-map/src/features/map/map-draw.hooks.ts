import { useCallback } from 'react'
import { useDispatch, useSelector } from 'react-redux'
import { selectDrawMode, setDrawMode, DrawMode } from './map.slice'

export const useMapDrawConnect = () => {
  const dispatch = useDispatch()
  const drawMode = useSelector(selectDrawMode)

  const dispatchSetDrawMode = useCallback(
    (mode: DrawMode) => {
      dispatch(setDrawMode(mode))
    },
    [dispatch]
  )

  return { drawMode, dispatchSetDrawMode }
}
