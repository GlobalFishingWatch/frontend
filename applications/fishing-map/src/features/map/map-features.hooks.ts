import { useEffect, useMemo, useState } from 'react'
import { atom, useRecoilState, useRecoilValue } from 'recoil'
import { TEMPORALGRID_SOURCE_LAYER, TimeChunks } from '@globalfishingwatch/layer-composer'
import {
  MERGED_ACTIVITY_ANIMATED_HEATMAP_GENERATOR_ID,
  UrlDataviewInstance,
} from '@globalfishingwatch/dataviews-client'
import { AnyGeneratorConfig } from '@globalfishingwatch/layer-composer/dist/generators/types'
import useMapInstance from 'features/map/map-context.hooks'
import { useGeneratorStyleMetadata, useMapStyle } from './map.hooks'

const sourcesLoadingState = atom<{ [key: string]: boolean }>({
  key: 'sourcesState',
  default: {},
  effects_UNSTABLE: [
    ({ onSet, setSelf, trigger, node }) => {
      const map = useMapInstance()

      // const sourceLoadingCallback = (sourcedataEvent: any) => {
      //   console.log('sourcedataloading')
      //   setSelf((s) => {
      //     console.log({ ...s, [sourcedataEvent.sourceId]: true })
      //     return { ...s, [sourcedataEvent.sourceId]: true }
      //   })
      // }
      // const sourceCallback = (sourcedataEvent: any) => {
      //   console.log('idle')
      //   // setSelf((s) => {
      //   //   return { ...s, [sourcedataEvent.sourceId]: true }
      //   // })
      // }
      // const sourceErrorCallback = (sourcedataEvent: any) => {
      //   console.log('err')
      //   // setSelf((s) => {
      //   //   return { ...s, [sourcedataEvent.sourceId]: true }
      //   // })
      // }
      if (map) {
        const sourceEventCallback = (e: any) => {
          const currentSources = Object.keys(map?.getStyle().sources || {})
          const sourcesLoaded = currentSources.map((sourceId) => [
            sourceId,
            map?.isSourceLoaded(sourceId),
          ])

          setSelf(Object.fromEntries(sourcesLoaded))
          // console.log(currentSources, sourcesLoaded)
        }
        map.on('sourcedataloading', sourceEventCallback)
        map.on('idle', sourceEventCallback)
        map.on('error', sourceEventCallback)
      }
      // console.log('onSet?', onSet, setSelf, trigger, map)
    },
  ],
})

export const useSourcesLoadingState = () => {
  const sourcesState = useRecoilValue(sourcesLoadingState)
  return sourcesState
}

export const useHaveAllSourcesLoaded = (sourcesIds: string[]) => {
  const sourcesState = useSourcesLoadingState()
  const haveAllSourcesLoaded = sourcesIds.every((sourceId) => sourcesState[sourceId] === true)
  return useMemo(() => {
    return haveAllSourcesLoaded
  }, [haveAllSourcesLoaded])
}

export const useFeatures = ({
  generators,
  sourceLayer = 'main',
  filter,
}: {
  generators: AnyGeneratorConfig[]
  sourceLayer?: string
  filter?: any[]
}) => {
  const sourcesMetadata = useActiveHeatmapAnimatedMetadatas(generators)
  const activeSourcesIds = useMemo(() => {
    return sourcesMetadata.map((metadata) => metadata?.timeChunks?.activeSourceId)
  }, [sourcesMetadata])
  const haveAllSourcesLoaded = useHaveAllSourcesLoaded(activeSourcesIds)
  const map = useMapInstance()

  const sourcesFeatures = useMemo(() => {
    if (haveAllSourcesLoaded && map) {
      const features = sourcesMetadata.map((metadata) => {
        const sourceId = metadata?.timeChunks?.activeSourceId
        const sourceFeatures = map.querySourceFeatures(sourceId, {
          sourceLayer,
          ...(filter && { filter }),
        })
        return sourceFeatures
      })
      return features
    }
  }, [haveAllSourcesLoaded, map, sourcesMetadata, sourceLayer, filter])

  return { sourcesFeatures, sourcesMetadata, haveAllSourcesLoaded }
}

export const useActiveHeatmapAnimatedMetadatas = (generators: AnyGeneratorConfig[]) => {
  const style = useMapStyle()
  const generatorsIds = generators.map((generator) => generator.id)
  const generatorsMetadata = generatorsIds.map((generatorId) => {
    return style?.metadata?.generatorsMetadata[generatorId]
  })
  const serializedGeneratorIds = generatorsMetadata
    .map((metadata) => {
      return metadata?.timeChunks?.activeSourceId
    })
    .join()
  const metadatas = useMemo(() => {
    return generatorsMetadata
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [serializedGeneratorIds])
  return metadatas
}

/**
TODO: DEPRECATE
 */
export const useMapSourceLoaded = (sourceId: string, cacheKey?: string) => {
  const map = useMapInstance()
  const [sourceLoaded, setSourceLoaded] = useState(false)

  // if cacheKey changed, reset loading status
  useEffect(() => {
    if (cacheKey) {
      setSourceLoaded(false)
    }
  }, [cacheKey])

  useEffect(() => {
    const sourceLoadingCallback = (sourcedataEvent: any) => {
      if (sourcedataEvent.sourceId === sourceId) {
        setSourceLoaded(false)
      }
    }

    // if no cachekey specified and sourceLoaded already true, start listen to sourceloading events
    // and reset loading status to false when source starts loading again
    // TODO could we simplify that by checking map?.isSourceLoaded() every time the hook is run?
    if (!cacheKey) {
      if (map && sourceLoaded) {
        map.on('sourcedataloading', sourceLoadingCallback)
      }
    }

    return () => {
      if (map) {
        map.off('sourcedataloading', sourceLoadingCallback)
      }
    }
  }, [cacheKey, map, sourceId, sourceLoaded])

  useEffect(() => {
    const sourceCallback = () => {
      if (map?.getSource(sourceId) && map?.isSourceLoaded(sourceId)) {
        setSourceLoaded(true)
        map.off('idle', sourceCallback)
      }
    }

    const sourceErrorCallback = (errorEvent: any) => {
      if (
        errorEvent.sourceId === sourceId &&
        map?.getSource(sourceId) &&
        map?.isSourceLoaded(sourceId)
      ) {
        setSourceLoaded(true)
      }
    }

    if (map && sourceId && !sourceLoaded) {
      map.on('idle', sourceCallback)
      map.on('error', sourceErrorCallback)
    }

    return () => {
      if (map) {
        map.off('idle', sourceCallback)
        map.off('error', sourceErrorCallback)
      }
    }
  }, [map, sourceId, sourceLoaded])

  return sourceLoaded
}

/**
TODO: DEPRECATE
 */
export const useMapFeatures = ({
  sourceId,
  sourceLayer = 'main',
  cacheKey,
  filter,
}: {
  sourceId: string
  sourceLayer?: string
  cacheKey?: string
  filter?: any[]
}) => {
  const map = useMapInstance()
  const sourceLoaded = useMapSourceLoaded(sourceId, cacheKey)
  // console.log(sourceId, sourceLayer, sourceLoaded, useMapStyle())

  const features = useMemo(() => {
    if (sourceLoaded) {
      const features = map?.querySourceFeatures(sourceId, {
        sourceLayer: sourceLayer,
        ...(filter && { filter }),
      })
      // console.log(features)
      return features
    }
  }, [sourceLoaded, map, sourceId, sourceLayer, filter])

  return { features, sourceLoaded }
}

/**
TODO: DEPRECATE
 */
const useMapTemporalgridLayerFeatures = ({
  cacheKey,
  generatorId,
}: {
  cacheKey?: string
  generatorId: string
}) => {
  const mergedActivityGenMetadata = useGeneratorStyleMetadata(generatorId)
  const currentTimeChunks = mergedActivityGenMetadata.timeChunks as TimeChunks
  const currentTimeChunkId = currentTimeChunks?.activeSourceId

  return useMapFeatures({
    sourceId: currentTimeChunkId,
    sourceLayer: TEMPORALGRID_SOURCE_LAYER,
    cacheKey,
  })
}

/**
TODO: DEPRECATE
 */
export const useActivityTemporalgridFeatures = ({ cacheKey }: { cacheKey?: string } = {}) => {
  return useMapTemporalgridLayerFeatures({
    cacheKey,
    generatorId: MERGED_ACTIVITY_ANIMATED_HEATMAP_GENERATOR_ID,
  })
}
