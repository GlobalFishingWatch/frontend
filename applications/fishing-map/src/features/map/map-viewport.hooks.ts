import { useSelector } from 'react-redux'
import { useCallback, useEffect, useState } from 'react'
import { ViewportProps } from 'react-map-gl'
import { MapCoordinates } from 'types'
import useDebounce from '@globalfishingwatch/react-hooks/dist/use-debounce'
import { selectViewport } from 'features/map/map.selectors'
import { useLocationConnect } from 'routes/routes.hook'

type SetMapCoordinatesArgs = Pick<ViewportProps, 'latitude' | 'longitude' | 'zoom'>
type UseViewport = {
  viewport: MapCoordinates
  onViewportChange: (viewport: ViewportProps) => void
  setMapCoordinates: (viewport: SetMapCoordinatesArgs) => void
}
export function useDebouncedViewport(
  urlViewport: MapCoordinates,
  callback: (viewport: MapCoordinates) => void
): UseViewport {
  const [viewport, setViewport] = useState<MapCoordinates>(urlViewport)
  const debouncedViewport = useDebounce<MapCoordinates>(viewport, 400)

  const setMapCoordinates = useCallback((viewport: SetMapCoordinatesArgs) => {
    setViewport({ ...viewport })
  }, [])

  const onViewportChange = useCallback((viewport: ViewportProps) => {
    const { latitude, longitude, zoom } = viewport
    setViewport({ latitude, longitude, zoom })
  }, [])

  // Updates local state when url changes
  useEffect(() => {
    const { latitude, longitude, zoom } = viewport
    if (
      urlViewport &&
      (urlViewport?.latitude !== latitude ||
        urlViewport?.longitude !== longitude ||
        urlViewport?.zoom !== zoom)
    ) {
      setViewport({ ...urlViewport })
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [urlViewport])

  // Sync the url with the local state debounced
  useEffect(() => {
    if (debouncedViewport && callback) {
      callback(debouncedViewport)
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [debouncedViewport])

  return { viewport, onViewportChange, setMapCoordinates }
}

export default function useViewport(): UseViewport {
  const { dispatchQueryParams } = useLocationConnect()
  const urlViewport = useSelector(selectViewport)
  const callback = useCallback((viewport) => dispatchQueryParams(viewport), [dispatchQueryParams])
  return useDebouncedViewport(urlViewport, callback)
}
