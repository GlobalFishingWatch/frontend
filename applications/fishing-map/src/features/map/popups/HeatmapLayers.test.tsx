import { Fragment, Suspense } from 'react'
import { render } from '@testing-library/react'
import { Provider } from 'react-redux'
import { Dataset } from '@globalfishingwatch/api-types/dist'
import datasets from 'features/datasets/datasets.mock'
import store from 'store'
import { TooltipEventFeature } from '../map.hooks'
import HeatmapTooltipRow from './HeatmapLayers'

jest.useFakeTimers()

describe('HeatmapTooltipRow', () => {
  const feature: TooltipEventFeature = {
    title: 'Some tooltip title',
    color: 'cyan',
    unit: 'meters',
    source: 'my unit test',
    sourceLayer: 'fake layer',
    layerId: 'some layer id',
    value: '7654',
    properties: { foo: 'bar' },
    vesselsInfo: {
      overflow: true,
      numVessels: 5,
      vessels: [
        {
          shipname: 'Grand Test',
          id: 'vessel-1',
          dataset: datasets.slice(0, 1).shift() as Dataset,
          hours: 12,
        },
        {
          shipname: 'Santa Partner',
          id: 'vessel-2',
          dataset: datasets.slice(0, 1).shift() as Dataset,
          hours: 20,
        },
      ],
    },
  }

  it('renders vessel info in tooltip', () => {
    const component = render(
      <Provider store={store}>
        <Suspense fallback={<Fragment />}>
          <HeatmapTooltipRow feature={feature} showFeaturesDetails={true} />
        </Suspense>
      </Provider>
    )
    jest.runAllTimers()
    expect(component.container).toMatchSnapshot()
  })
})
