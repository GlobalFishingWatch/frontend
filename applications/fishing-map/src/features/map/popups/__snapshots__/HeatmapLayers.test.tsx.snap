// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HeatmapTooltipRow renders vessel info in tooltip 1`] = `
<div>
  <div
    class="popupSection"
  >
    <span
      class="popupSectionColor"
      style="background-color: cyan;"
    />
    <div
      class="popupSectionContent"
    >
      <h3
        class="popupSectionTitle"
      >
        Some tooltip title
      </h3>
      <div
        class="row"
      >
        <span
          class="rowText"
        >
          7,654
           
          common.meters
          common.hour
        </span>
      </div>
      <div
        class="vesselsTable"
      >
        <div
          class="vesselsHeader"
        >
          <label
            class="vesselsHeaderLabel"
          >
            common.vessel_plural
          </label>
          <label
            class="vesselsHeaderLabel"
          >
            common.hour_plural
          </label>
        </div>
        <button
          class="vesselRow"
        >
          <span
            class="vesselName"
          >
            Grand Test
            <span
              class="vesselRowLegend"
            >
               - 
              Port Events (AIS)
            </span>
          </span>
          <span>
            12
          </span>
        </button>
        <button
          class="vesselRow"
        >
          <span
            class="vesselName"
          >
            Santa Partner
            <span
              class="vesselRowLegend"
            >
               - 
              Port Events (AIS)
            </span>
          </span>
          <span>
            20
          </span>
        </button>
        <div
          class="vesselsMore"
        >
          + 
          3
           
          common.more
        </div>
      </div>
    </div>
  </div>
</div>
`;
